### PR TITLE
feat(plugin-jwt) support RS512 algorithm for signing tokens.

### PR DESCRIPTION
**Pull request by sarraz1**
Add support of RS512 algorithm for signing tokens in jwt plugin.

This add-on is based on the already-used module "crypto".

### CODE REVIEW

File: kong/plugins/jwt/jwt_parser.lua [hunk 1]
```diff
@@ -49,6 +50,10 @@ local alg_verify = {
     local pkey = assert(crypto.pkey.from_pem(key), "Consumer Public Key is Invalid")
     return crypto.verify('sha256', data, signature, pkey)
   end,
+  ["RS512"] = function(data, signature, key)
+     local pkey = assert(crypto.pkey.from_pem(key), "Consumer Public Key is Invalid")
```
Comment (p0pr0ck5) [inline kong/plugins/jwt/jwt_parser.lua]: please check indentation spacing here as well.

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 2]
```diff
@@ -76,7 +76,29 @@ describe("Plugin: jwt (parser)", function()
       local token = jwt_parser.encode({sub = "1234"}, "secret", nil, {alg = "RS256"})
       assert(jwt_parser:new(token))
     end)
+
+    it("accepts a valid RS512 encoding request", function()
+      local token = jwt_parser.encode({sub = "1234"}, "secret", nil, {
+        typ = "JWT",
+        alg = "RS512"
```
Comment (p0pr0ck5) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: minor style nitpick, its preferred to leave a trailing comma after the last item in a table to facilitate cleaner future commit histories.

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 3]
```diff
@@ -76,7 +76,29 @@ describe("Plugin: jwt (parser)", function()
...
+      })
+      assert(jwt_parser:new(token))
+    end)
+    it("accepts a valid RS512 encoding request with lowercase TYP", function()
+      local token = jwt_parser.encode({sub = "1234"}, "secret", nil, {
+        typ = "jwt",
+        alg = "RS512"
```
Comment (p0pr0ck5) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: same as above.

File: spec/03-plugins/17-jwt/03-access_spec.lua [hunk 4]
```diff
@@ -267,6 +273,41 @@ describe("Plugin: jwt (access)", function()
...
+    it("verifies JWT", function()
+      PAYLOAD.iss = rsa_jwt_secret_3.key
+      local jwt = jwt_encoder.encode(PAYLOAD, fixtures.rs512_private_key, 'RS512')
+      local authorization = "Bearer "..jwt
+      local res = assert(proxy_client:send {
+        method = "GET",
+        path = "/request",
+        headers = {
+          ["Authorization"] = authorization,
+          ["Host"] = "jwt.com"
```
Comment (p0pr0ck5) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: trailing coma here

File: spec/03-plugins/17-jwt/03-access_spec.lua [hunk 5]
```diff
@@ -267,6 +273,41 @@ describe("Plugin: jwt (access)", function()
...
+    it("identifies Consumer", function()
+      PAYLOAD.iss = rsa_jwt_secret_3.key
+      local jwt = jwt_encoder.encode(PAYLOAD, fixtures.rs512_private_key, 'RS512')
+      local authorization = "Bearer "..jwt
+      local res = assert(proxy_client:send {
+        method = "GET",
+        path = "/request",
+        headers = {
+          ["Authorization"] = authorization,
+          ["Host"] = "jwt.com"
```
Comment (p0pr0ck5) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: trailing coma here

**Comment by p0pr0ck5** [comment]
@sarraz1 any update based on the comments above?

**Comment by sarraz1** [comment]
I  will  submit on master with every required updates soon (not today).

File: kong/plugins/jwt/daos.lua [hunk 6]
```diff
@@ -10,9 +10,9 @@ local SCHEMA = {
     created_at = {type = "timestamp", immutable = true, dao_insert_value = true},
     consumer_id = {type = "id", required = true, foreign = "consumers:id"},
     key = {type = "string", unique = true, default = utils.random_string},
-    secret = {type = "string", default = utils.random_string},
+    secret = {type = "string", unique = true, default = utils.random_string},
```
Comment (sarraz1) [inline kong/plugins/jwt/daos.lua]: there is no reason.

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua (hunk 2 above)
Comment (sarraz1) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: so:
alg = "RS512",

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 7]
```diff
@@ -90,6 +112,16 @@ describe("Plugin: jwt (parser)", function()
...
       assert.False(jwt:verify_signature(fixtures.rs256_public_key:gsub('QAB', 'zzz')))
     end)
+    it("using RS512", function()
+      local token = jwt_parser.encode({sub = "foo"}, fixtures.rs512_private_key, 'RS512')
+      local jwt = assert(jwt_parser:new(token))
+      assert.True(jwt:verify_signature(fixtures.rs512_public_key))
+      assert.False(jwt:verify_signature(fixtures.rs512_public_key:gsub('AE=', 'zzz')))
+    end)
+    -- it("test", function()
+    --assert.True(TRUE)
```
Comment (sarraz1) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: ok

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua (hunk 3 above)
Comment (sarraz1) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: yes

**Comment by sarraz1** [comment]
Like you suggested, I submitted a new PR from master: https://github.com/Mashape/kong/pull/2666 
So you can cancel this one.

Regards

**Comment by thibaultcha** [comment]
Closing as this is replaced by the new PR.

**Pull request closed without merging**